crates/lib/src/qvm/mod.rs

```diff
@@ -32,73 +41,125 @@ impl QvmResultData {
     }
 }
 
-#[derive(Debug, Deserialize, Clone, PartialEq)]
-#[serde(untagged)]
-pub(super) enum Response {
-    Success(Success),
-    Failure(Failure),
-}
-
-#[derive(Debug, Deserialize, Clone, PartialEq)]
-pub(super) struct Success {
-    #[serde(flatten)]
-    pub(super) registers: HashMap<String, RegisterData>,
-}
-
-#[derive(Debug, Deserialize, Clone, Eq, PartialEq)]
-pub(super) struct Failure {
-    /// The message from QVM describing what went wrong.
-    pub(super) status: String,
+/// Run a Quil program on the QVM. The given [`Parameters`] are used to parameterize the value of
+/// memory locations across shots.
+#[allow(clippy::too_many_arguments)]
+pub async fn run(
+    quil: &str,
+    shots: u16,
+    addresses: HashMap<String, AddressRequest>,
+    params: &Parameters,
+    measurement_noise: Option<(f64, f64, f64)>,
+    gate_noise: Option<(f64, f64, f64)>,
+    rng_seed: Option<i64>,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!("parsing a program to be executed on the qvm");
+    let program = Program::from_str(quil).map_err(Error::Parsing)?;
+    run_program(
+        &program,
+        shots,
+        addresses,
+        params,
+        measurement_noise,
+        gate_noise,
+        rng_seed,
+        config,
+    )
+    .await
 }
 
-#[derive(Serialize, Debug, Clone, PartialEq, Eq)]
-#[serde(rename_all = "kebab-case")]
-struct Request<'request> {
-    quil_instructions: String,
-    addresses: HashMap<&'request str, bool>,
-    trials: u16,
-    #[serde(rename = "type")]
-    request_type: RequestType,
-}
+/// Run a [`Program`] on the QVM. The given [`Parameters`] are used to parametrize the value of
+/// memory locations across shots.
+#[allow(clippy::too_many_arguments)]
+pub async fn run_program(
+    program: &Program,
+    shots: u16,
+    addresses: HashMap<String, AddressRequest>,
+    params: &Parameters,
+    measurement_noise: Option<(f64, f64, f64)>,
+    gate_noise: Option<(f64, f64, f64)>,
+    rng_seed: Option<i64>,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!(
+        %shots,
+        ?addresses,
+        ?params,
+        "executing program on QVM"
+    );
+    if shots == 0 {
+        return Err(Error::ShotsMustBePositive);
+    }
```

Comment:
https://doc.rust-lang.org/stable/std/num/struct.NonZeroU16.html

Reply:
None of this is new code, should have created an issue, but I went ahead and refactored the type used for shots to be `NonZeroU16`. Ended up impacting a lot of the crate.